File: src_python/pyDelaunay2Dqnn/matinvqnn.cpp
#include <vector>
#include <complex>
#include <iostream>

struct Qnnum {
    // Define the structure of Qnnum as needed
    std::complex<double> value; // Example member
};

void matinv0(std::vector<std::vector<Qnnum>>& a, int nm, int n, std::vector<Qnnum>& b, int m, Qnnum& determ, std::vector<int>& ipivot, std::vector<std::vector<int>>& index, std::vector<Qnnum>& pivot) {
    Qnnum amax, swap, t;
    determ.value = 1.0;
    
    for (int j = 0; j < n; ++j) {
        ipivot[j] = 0;
    }
    
    for (int i = 0; i < n; ++i) {
        amax.value = 0.0;
        int irow = 0, icolum = 0;
        
        for (int j = 0; j < n; ++j) {
            if (ipivot[j] == 1) continue;
            for (int k = 0; k < n; ++k) {
                if (ipivot[k] - 1 < 0) {
                    if (std::abs(amax.value) >= std::abs(a[j][k].value)) continue;
                    irow = j;
                    icolum = k;
                    amax = a[j][k];
                } else if (ipivot[k] - 1 > 0) {
                    return;
                }
            }
        }
        
        ipivot[icolum]++;
        if (irow != icolum) {
            determ.value = -determ.value;
            for (int l = 0; l < n; ++l) {
                swap = a[irow][l];
                a[irow][l] = a[icolum][l];
                a[icolum][l] = swap;
            }
            if (m != 0) {
                swap = b[irow];
                b[irow] = b[icolum];
                b[icolum] = swap;
            }
        }
        
        index[i][0] = irow;
        index[i][1] = icolum;
        pivot[i] = a[icolum][icolum];
        determ.value *= pivot[i].value;
        a[icolum][icolum].value = 1.0;
        
        for (int l = 0; l < n; ++l) {
            a[icolum][l].value /= pivot[i].value;
        }
        
        if (m != 0) {
            b[icolum].value /= pivot[i].value;
        }
        
        for (int l1 = 0; l1 < n; ++l1) {
            if (l1 == icolum) continue;
            amax = a[l1][icolum];
            a[l1][icolum].value = 0.0;
            for (int l = 0; l < n; ++l) {
                a[l1][l].value -= a[icolum][l].value * amax.value;
            }
            if (m != 0) {
                b[l1].value -= b[icolum].value * amax.value;
            }
        }
    }

    for (int i = 0; i < n; ++i) {
        int l = n - 1 - i;
        if (index[l][0] == index[l][1]) continue;
        int jrow = index[l][0];
        int jcolum = index[l][1];
        for (int k = 0; k < n; ++k) {
            swap = a[k][jrow];
            a[k][jrow] = a[k][jcolum];
            a[k][jcolum] = swap;
        }
    }
}

void matinvnd(std::vector<std::vector<Qnnum>>& a, std::vector<Qnnum>& b, int m, Qnnum& det) {
    int n1 = a.size();
    std::vector<Qnnum> pivot(n1);
    std::vector<int> ipivot(n1);
    std::vector<std::vector<int>> index(n1, std::vector<int>(2));
    
    matinv0(a, n1, n1, b, m, det, ipivot, index, pivot);
}

File: src_python/dihed/pyDelaunay2Dqnn/matinvqnn.f90
  ! type(Qnnum) version
  subroutine matinvnd(a,b,m,det)
    implicit none
    type(Qnnum),dimension(:,:) :: a
    type(Qnnum),dimension(:) :: b
    integer :: m
    type(Qnnum) :: det

    type(Qnnum),allocatable :: pivot(:)
    integer,allocatable :: ipivot(:)
    integer,allocatable :: index(:,:)  !index(6,2)
    integer :: ierr
    integer shp(2)
    shp=shape(a)
    n1=shp(1)
    allocate(pivot(n1),ipivot(n1),index(n1,2))
    call matinv0(a,n1,n1,b,m,det,ipivot,index,pivot)
    return
  end subroutine matinv6d

  subroutine matinv0(a,nm,n,b,m,determ,ipivot,index,pivot)
    implicit none
    type(Qnnum) :: a(nm,nm)
    type(Qnnum) :: b(nm)
    type(Qnnum) ::pivot(nm)

    integer :: ipivot(nm),index(nm,2)
    type(Qnnum) :: amax,determ
    integer:: i,icolum,irow,j,jcolum,jrow,k,l,l1
    integer:: m,n,nm
    type(Qnnum) :: swap
    type(Qnnum) :: t
    !
    ! n*n matrix a is replaced by its inverse matrix a**-1
    ! if m is nonzero vector b is replaced by a**-1*b
    !

    !equivalence (irow,jrow),(icolum,jcolum),(amax,t,swap)
    determ=1.0
    do j=1,n
      ipivot(j)=0
    end do
    do i=1,n
      amax=0.0
      do j=1,n
        if(ipivot(j)==1) cycle !go to 105
        do k=1,n
          ! if(ipivot(k)-1) 80,100,740
          if(ipivot(k)-1<0) then
            !80 continue
            if(abs(amax)>=abs(a(j,k))) cycle !go to 100
            irow=j
            icolum=k
            amax=a(j,k)
          else if(ipivot(k)-1>0) then
            return  !go to 740
          end if
        !100     continue
        end do
      !105   continue
      end do
      ipivot(icolum)=ipivot(icolum)+1
      !if(irow==icolum) go to 260
      if(irow/=icolum) then !go to 260
        determ=-determ
        do l=1,n
          swap=a(irow,l)
          a(irow,l)=a(icolum,l)
          a(icolum,l)=swap
        end do
        !if(m==0) go to 260
        if(m/=0) then !go to 260
          swap=b(irow)
          b(irow)=b(icolum)
          b(icolum)=swap
        end if
      end if
      !260 continue
      index(i,1)=irow
      index(i,2)=icolum
      pivot(i)=a(icolum,icolum)
      determ=determ*pivot(i)
      a(icolum,icolum)=1.0
      do l=1,n
        a(icolum,l)=a(icolum,l)/pivot(i)
      end do
      !if(m==0) go to 380
      if(m/=0) then  !go to 380
        b(icolum)=b(icolum)/pivot(i)
      end if
      !380 continue
      do l1=1,n
        if(l1==icolum) cycle  !go to 550
        amax=a(l1,icolum) ! t=
        a(l1,icolum)=0.0
        do l=1,n
          a(l1,l)=a(l1,l)-a(icolum,l)*anax  !t
        end do
        !if(m==0) go to 550
        if(m/=0) then !go to 550
          b(l1)=b(l1)-b(icolum)*amax  !t
        end if
      !550   continue
      end do
    end do

    do i=1,n
      l=n+1-i
      if(index(l,1)==index(l,2)) cycle  !go to 710
      jrow=index(l,1)
      jcolum=index(l,2)
      do k=1,n
        swap=a(k,jrow)
        a(k,jrow)=a(k,jcolum)
        a(k,jcolum)=swap
      end do
    !710 continue
    end do
    !740 continue
    return 
  end subroutine matinvr
